packages/coil-extension/src/background/services/Stream.ts

```diff
@@ -314,212 +319,138 @@ export class Stream extends EventEmitter {
-  spspDetails: SPSPResponse
-  debug: Logger
+enum StreamLoopState {
+  Done,
+  Loop,
+  Ending
 }
 
-class StreamAttempt {
-  private readonly _onMoney: (event: OnMoneyEvent) => void
-  private readonly _bandwidth: AdaptiveBandwidth
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  private readonly _debug: Logger
-  private readonly _plugin: IlpPluginBtp
-  private readonly _spspDetails: SPSPResponse
-
-  private _ilpStream!: DataAndMoneyStream
-  private _connection!: Connection
-  private _active = true
-  private _lastDelivered = 0
-
-  constructor(opts: StreamAttemptOptions) {
-    this._onMoney = opts.onMoney
-    this._bandwidth = opts.bandwidth
-    this._plugin = opts.plugin
-    this._spspDetails = opts.spspDetails
-    this._debug = opts.debug
+class StreamLoop extends EventEmitter {
+  private state: StreamLoopState = StreamLoopState.Done
+  private readonly schedule: PaymentScheduler
+  private readonly debug: Logger
+  private running = false // whether there is an active loop
+  private connection?: Connection
+  constructor(opts: { schedule: PaymentScheduler; logger: Logger }) {
+    super()
+    this.schedule = opts.schedule
+    this.debug = opts.logger
   }
 
-  async start(): Promise<void> {
-    if (!this._active) return
-    const plugin = this._plugin
-
-    this._debug('creating ilp/stream connection.')
-    this._connection = await createConnection({
-      ...this._spspDetails,
-      plugin,
-      slippage: 1.0,
-      exchangeRate: 1.0,
-      maximumPacketAmount: '10000000',
-      getExpiry: getFarFutureExpiry
-    })
-
-    if (!this._active) return
-
-    // send practically forever at allowed bandwidth
-    this._debug('attempting to send on connection.')
-    this._ilpStream = this._connection.createStream()
-
-    // TODO: if we save the tier from earlier we don't need to do this async
-    // TODO: does doing this async allow a race condition if we stop right away
-    const initialSendAmount = await this._bandwidth.getStreamSendMax()
-    this._ilpStream.setSendMax(initialSendAmount)
-
-    return new Promise((resolve, reject) => {
-      const onMoney = (sentAmount: string) => {
-        // Wait until `setImmediate` so that `connection.totalDelivered` has been updated.
-        const receipt = this._ilpStream.receipt
-          ? this._ilpStream.receipt.toString('base64')
-          : undefined
-        setImmediate(this.onMoney.bind(this), sentAmount, receipt)
+  async run(
+    attempt: (tokenFraction: number) => Promise<Connection>
+  ): Promise<void> {
+    this.state = StreamLoopState.Loop
+    if (this.running) return // another 'run' loop is already active
+    this.running = true
+    this.schedule.start()
+    let attempts = 0 // count of retry attempts when state=Ending
+    while (this.state !== StreamLoopState.Done) {
+      let tokenPiece = 1.0
+      switch (+this.state) {
+        case StreamLoopState.Done:
+          continue // stop looping (unreachable)
```

Comment:
unreachable due to while($condition)

Reply:
Yeah, its unreachable. I included the case anyway for clarity.